plugins/global-search/src/utils/indexer/indexer.ts

```diff
@@ -0,0 +1,183 @@
+import { type IndexEntry, includedAttributes, type RootNode, shouldIndexNode } from "./types"
+
+async function getNodeName(node: AnyNode): Promise<string | null> {
+    if (isWebPageNode(node)) {
+        if (node.path === "/") return "Home"
+        if (!node.path) return "Web Page"
+        return node.path
+    }
+
+    if (isComponentNode(node)) {
+        return node.name ?? "Component"
+    }
+
+    if (isFrameNode(node)) {
+        return node.name ?? "Frame"
+    }
+
+    if (isTextNode(node)) {
+        return (await node.getText()) ?? null
+    }
+
+    return null
+}
+
+export class GlobalSearchIndexer {
+    private eventEmitter = new TypedEventEmitter()
+    public on: TypedEventEmitter["on"] = (...args) => this.eventEmitter.on(...args)
+
+    private entries: Record<string, IndexEntry> = {}
+    private batchSize = 10
+
+    private abortRequested = false
+
+    private upsertEntries(entries: IndexEntry[]) {
+        for (const entry of entries) {
+            this.entries[entry.id] = entry
+            this.eventEmitter.emit("upsert", { entry })
+        }
+    }
+
+    private async getNodeText(node: AnyNode): Promise<string | null> {
+        if (includedAttributes.includes("text") && isTextNode(node)) {
+            return await node.getText()
+        }
+        return null
+    }
+
+    private async *crawlNodes(rootNodes: readonly RootNode[]): AsyncGenerator<IndexEntry[]> {
+        let batch: IndexEntry[] = []
+
+        for (const rootNode of rootNodes) {
+            const rootNodeName = await getNodeName(rootNode)
+
+            for await (const node of rootNode.walk()) {
+                if (this.abortRequested) return
+
+                if (!shouldIndexNode(node)) continue
+
+                const [text, name] = await Promise.all([this.getNodeText(node), getNodeName(node)])
+
+                batch.push({
+                    id: node.id,
+                    type: node.__class,
+                    name,
+                    text,
+                    node,
+                    rootNode,
+                    rootNodeName,
+                    rootNodeType: rootNode.__class,
+                })
+
+                if (batch.length === this.batchSize) {
+                    yield batch
+                    batch = []
+                }
+            }
+        }
+
+        if (batch.length > 0) {
+            yield batch
+        }
+    }
+
+    private async *crawlCollections(collections: readonly Collection[]): AsyncGenerator<IndexEntry[]> {
+        let batch: IndexEntry[] = []
+
+        for (const collection of collections) {
+            const [fieldNames, items] = await Promise.all([collection.getFields(), collection.getItems()])
+            const fieldNameMap = new Map(fieldNames.map(f => [f.id, f.name]))
+
+            for (const item of items) {
+                const fields = Object.fromEntries(
+                    Object.entries(item.fieldData).flatMap(([key, field]) => {
```

Comment:
nit: I find this quite hard to read with the nesting. I feel like a regular `for ... of` loop would be more readable and consistent with code style in FramerStudio (but am nitpicking so feel free to ignore)

Reply:
🙉 🙈 🙊